Adds page objects and step definitions for navigating around the WordPress admin

stephenharris opened the pull request:
## Description

Adds the following `Page`/`Element` classes:

 - `AdminPage` - a non-specific admin page. Acts as base for specific admin pages, but useful its own right for step definitions which are not page-specific.
 - `Dashboard` (i.e. `wp-admin/index.php`).
 - `AdminMenu` - an `Element` which allows you to list all menu items or click an specifc menu item
 
Adds the following steps:

 - `Then I should be on the <page title> page`
 - `Given I am on the Dashboard`
 - `Given I go to menu item :item` (e.g. `Given I go to Posts > Add New`)
 - `When I click on the :link link in the header` (e.g `When I click on the Add New link in the header`) \*

\* This fails on WordPress' nightly build. It appears links have been removed from the header, so the implementation of this step would need to be updated accordingly.

Fixes:

 - Admin password in `behat.yml` was incorrect.

## Related issue
Ref #26

## Motivation and context
The aim of Page and Element objects is to encapsulate knowledge of the page and element mark-up, so that Contexts can interact with pages, or specific parts of pages, without knowing details about them. This keeps contexts clean and makes maintaining the WordHat easier (as the impact of changes to WordPress' markup will be limited to Page and Element objects).

While the scope for WordHat would be limited to the admin side there is no reason why the end user couldn't create their own pages and elements for the front-end.

## How has this been tested?
Yes, via `features/user.feature`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

paulgibbs commented on src/PageObject/Element/AdminMenu.php:
```diff
@@ -0,0 +1,108 @@
+     */
+    public function getTopLevelMenuItems()
+    {
+        $menuItem_nodes = $this->findAll('css', '#adminmenu > li a .wp-menu-name');
```
There may be further instances but for variable names please can we use underscore seperated , not camel case? We're only using camelcase for method names per PSR1. Thank you.

paulgibbs replied:
I've pointed this out here but I can see this occurs in other part of the changeset

paulgibbs commented on src/PageObject/Element/AdminMenu.php:
```diff
@@ -0,0 +1,108 @@
+     * @param string $html Raw HTML
+     * @param string Extracted text. e.g. Some <span>HTML and</span>text  -->  Some text
+     */
+    private function stripTagsAndContent($html)
```
Open for discussion but I dislike private visibility methods because if i'm another developer and i'm trying to hack around/extend something, it's super annoying to come across a method which is private. Can we use protected instead, please?

stephenharris commented on features/user.feature:
```diff
@@ -0,0 +1,24 @@
+		And I am on the Dashboard
+
+  @javascript
+  Scenario: I can add a new user
```
The second one yes, because selecting a submenu item has to be done differently when you are using a javascript enabled driver (because the submenu item is not immediately visible).

paulgibbs replied:
OK!

stephenharris replied:
(Actually they both do - so I've reinstated it)

paulgibbs replied:
also ok!

stephenharris commented:
Sorry - I had two local copies (different machines), and things got out of sync. The most straightforward solution was to force an update and overwrite the history. The commits associated to your comments no longer exist, but hopefully it's clear that I've addressed the issues raised.

paulgibbs commented:
Will take another look!

paulgibbs commented on features/bootstrap/FeatureContext.php:
```diff
@@ -1,20 +0,0 @@
-<?php
```
Please can we reinstate this file?

The reason being, on https://wordhat.info/installation/index.html on step 3 and 4, `vendor/behat --init` throws a warning and does not create `/features/bootstrap/featurecontext.php`  unless this file existed

paulgibbs commented on src/Context/RawWordpressContext.php:
```diff
@@ -163,7 +162,7 @@ public function logIn($username, $password)
         }
 
-        $this->visitPath('wp-login.php?redirect_to=' . urlencode($this->locatePath('/')));
+        $this->visitPath('/wp-login.php?redirect_to=' . urlencode($this->locatePath('/')));
```
Does this need the `/` added? I think `locatePath` will sort this one.

paulgibbs commented on src/PageObject/AdminPage.php:
```diff
@@ -0,0 +1,148 @@
+    {
+        $actual = $this->getHeaderText();
+        if ($expected !== $actual) {
+            throw new \Exception(sprintf('Expected page header "%s", found "%s".', $expected, $actual));
```
I think we should use `ExpectationException` if we're trying to assert that something should occur on the webpage.

See http://www.tentacode.net/10-tips-with-behat-and-mink step 4.

paulgibbs commented on src/PageObject/AdminPage.php:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @throws \Exception
+     */
+    private function getHeaderElement()
```
protected please

stephenharris replied:
Where do we stand on breaking changes? The advantage of `private` is that we know we can change or remove the function without breaking anything. If someone needs access to it, they can copy it into their own code. We're still at the alpha stage so it's not an issue - but if we want to avoid breaking changes going forward we might need to think critically about the API we are exposing.

paulgibbs replied:
We should have a ticket open to discuss this!

When I start WordHat I figured it'd be a hoot to try using Composer and semver, along with PSR2/4 compliance, etc -- to end up with better? code that was more native PHP than "WordPress"-ish. Totally up for anything more interesting you want to try -- including calling it "Version 1", "Version 2" and just incrementing one digit forever.

paulgibbs replied:
Right now I only know of a few? of people using this for anything meaningful which is why you see me pinging Dan sometimes, because I've an idea of what he's done. Doesn't scale, obvs.

stephenharris replied:
I'm definitely in favour of semver. I do prefer to avoid breaking changes until the change is something significant, and not just because a single (public/protected) function was changed which is why I generally err on making methods private. That said, if we're happy to make breaking changes then that's fine - I just didn't want decisions made this early on to hamper future iterations.

paulgibbs commented on src/PageObject/AdminPage.php:
```diff
@@ -0,0 +1,148 @@
+     * @see https://github.com/sensiolabs/BehatPageObjectExtension/issues/57
+     * @param array $url_parameters
+     * @return boolean
+     * @throws SensioLabs\Behat\PageObjectExtension\PageObject\Exception\UnexpectedPageException
```
FYI I think `@throws` is one of the more useless PHPDoc tags, so I don't mind having them, but i'm not planning to use it in my own future PRs.

stephenharris replied:
Why useless? Unless there is a throw explicitly in the method then there is no obvious indication that it would throw anything. Even if it is there, I wouldn't regard it as any less useless than the method's `@return` - they're both part of the 'contract'.

paulgibbs replied:
I think if I wanted to put a note for a Human, i'd put it in the function's long description. Putting stuff in these at-tags makes it easier for PHPDoc to extract that, and do its work. As far as I'm concerned, that kind of document parser/IDE ought to be smart enough to analyse the code to see if an exception is thrown, or have enough context that it knows a called function can throw an exception -- and with that innate understanding, it can go build what output it wants.

Just my 5 cents

paulgibbs commented on src/PageObject/AdminPage.php:
```diff
@@ -0,0 +1,148 @@
+     */
+    private function makeSurePathIsAbsolute($path)
+    {
+        $site_url = rtrim($this->getParameter('site_url'), '/').'/';
```
Can we have spaces around the `.` concatenation? Here and on the line below.

paulgibbs commented on src/PageObject/Element/AdminMenu.php:
```diff
@@ -0,0 +1,127 @@
+     */
+    protected function stripTagsAndContent($html)
+    {
+        if (trim($html) == '') {
```
Strict comparison operators please

paulgibbs commented:
PR is OK to merge but have left some comments that'll help make it perfect.
Sorry to be so picky -- not intending to annoy or discourage you. Hopefully when we get some more PHPCS or PHPMD pre-commit hooks in place, this will become less common

stephenharris commented:
No worries :), and thanks for the heads up on `ExpectationException` - I wasn't aware of it. 

I've a `phpcs` rule for checking for spaces around concatenation. I wanted to add one for strict operators, but `Squiz.Operators.ComparisonOperatorUsage` also flags where `!...` is used instead of `...=== FALSE`, but `isset(...) === false` and `is_empty(...) === false`.

paulgibbs commented:
Going to merge this! Thanks for all the work! Excited to have in.
